ui/app/components/http-requests-bar-chart.js

```diff
@@ -0,0 +1,206 @@
+    total: 50000,
+  },
+  {
+    start_time: '2019-04-01T00:00:00Z',
+    total: 4500,
+  },
+  {
+    start_time: '2019-03-01T00:00:00Z',
+    total: 550000,
+  },
+];
+
+export default Component.extend({
+  classNames: ['http-requests-bar-chart-container'],
+  counters: COUNTERS,
+  svgContainer: null,
+  barsContainer: null,
+  clipPathRect: null,
+  margin: { top: 24, right: 16, bottom: 24, left: 16 },
+  width: 0,
+  height() {
+    const { margin } = this;
+    return 240 - margin.top - margin.bottom;
+  },
+
+  parsedCounters: computed('counters', function() {
+    // parse the start times so bars and ticks display properly
+    const { counters } = this;
+    return counters.map(counter => {
+      return assign({}, counter, { start_time: d3TimeFormat.isoParse(counter.start_time) });
+    });
+  }),
+
+  yScale: computed('parsedCounters', 'height', function() {
+    const { parsedCounters } = this;
+    const height = this.height();
+    const counterTotals = parsedCounters.map(c => c.total);
+
+    return d3Scale
+      .scaleLinear()
+      .domain([0, Math.max(...counterTotals)])
+      .range([height, 0]);
+  }),
+
+  xScale: computed('parsedCounters', 'width', function() {
+    const { parsedCounters, width, margin } = this;
+
+    return d3Scale
+      .scaleBand()
+      .domain(parsedCounters.map(c => c.start_time))
+      .rangeRound([0, width - margin.left - margin.right], 0.05)
+      .paddingInner(0.04);
+  }),
+
+  didInsertElement() {
+    this._super(...arguments);
+    const { margin } = this;
+
+    // set the width after the element has been rendered because the chart axes depend on it.
+    // this helps us avoid an arbitrary hardcoded width which causes alignment & resizing problems.
+    run.schedule('afterRender', this, () => {
+      this.set('width', this.element.clientWidth - margin.left - margin.right);
+      this.initBarChart();
+    });
+  },
+
+  initBarChart() {
+    // mount the d3 elements on initial page load, but do not render the bars yet.
+    const svgContainer = d3.select('.http-requests-bar-chart');
+    this.set('svgContainer', svgContainer);
+
+    const defs = svgContainer.append('defs');
+
+    const bgGradient = defs
+      .append('linearGradient')
+      .attr('id', 'bg-gradient')
+      .attr('gradientTransform', 'rotate(90)');
+
+    bgGradient
+      .append('stop')
+      // this corresponds to $blue-500
+      .attr('stop-color', '#1563ff')
+      .attr('stop-opacity', '0.8')
+      .attr('offset', '0%');
+
+    bgGradient
+      .append('stop')
+      // this corresponds to $blue-500
+      .attr('stop-color', '#1563ff')
+      .attr('stop-opacity', '0.3')
+      .attr('offset', '100%');
+
+    const clipPathRect = svgContainer
+      .append('g')
+      .attr('clip-path', `url(#clip-bar-rects)`)
+      .append('rect')
+      .attr('x', 0)
+      .attr('y', 0)
+      .style('fill', 'url(#bg-gradient)');
+
+    this.set('clipPathRect', clipPathRect);
```

Comment:
What do you think about moving defs and clipPath stuff to the SVG in the template? These seem more declarative, and I don't think they're getting updated on re-render.

Reply:
ohhh yess, this is an excellent idea. it'll make this method MUCH more readable. thank you for the suggestion!